main/main.c

```diff
@@ -484,6 +491,66 @@ static void batchMakeTags (cookedArgs *args, void *user CTAGS_ATTR_UNUSED)
 #undef timeStamp
 }
 
+#ifdef HAVE_JANSSON
+void interactiveLoop (cookedArgs *args, void *user CTAGS_ATTR_UNUSED)
+{
+  char buffer[1024];
+  json_t *request;
+
+  fprintf (stdout, "{\"name\": \"" PROGRAM_NAME "\", \"version\": \"" PROGRAM_VERSION "\"}\n");
+  fflush (stdout);
+
+  while (fgets (buffer, sizeof(buffer), stdin)) {
+    if (buffer[0] == '\n')
+      continue;
+
+    request = json_loads (buffer, JSON_DISABLE_EOF_CHECK, NULL);
+    if (! request) {
+      error (FATAL, "invalid json");
+      goto next;
+    }
+
+    json_t *command = json_object_get (request, "command");
+    if (! command) {
+      error (FATAL, "command name not found");
+      goto next;
+    }
+
+    if (!strcmp ("list-parsers", json_string_value (command))) {
+            /* todo */
+    } else if (!strcmp ("generate-tags", json_string_value (command))) {
+      json_int_t size = -1;
+      const char *filename;
+
+      if (json_unpack (request, "{s?I ss}", "size", &size, "filename", &filename) == -1) {
```

Comment:
The `?` syntax means optional key, and is new to jansson 2.8. Since some distributions do not have this version available, I will need to find a backwards-compatible way to achieve the same result.